EIPS/eip-3332.md

```diff
@@ -0,0 +1,95 @@
+
+## Motivation
+
+With the emergence of rollups as core mechanisms in scaling Ethereum there are a number of common transactions that can be front-run. Optimistic rollups rely on the submission of fraud proofs to maintain the integrity of their systems. As a result actors submitting fraud proofs typically receive a financial reward for doing so. This opens a trivial front-running strategy of watching the mempool for fraud proof submissions and copying such transactions with a much higher gas price to reap the reward. Such front-runners do not perform validation independently and de-incentivize others from performing validation. Adding a mechanism enforcing an upper bound on gas prices for a transaction could be an effective defense against such front-running attacks.
+
+## Specification
+
+If `block.number >= TBD`, add a new opcode:
+
+MEDGASPRICE: 0x46
+
+Pushes the median gas price of the parent block onto the stack.<br />
+Gas costs: 8
+
+## Rationale
+
+Consider a smart contract that wants to implement a first come first serve mechanism. Such a mechanism must defeat the inherently pay-to-win nature of the gas price market. Enforcing a maximum gas price for a transaction relies on the fact that transactions of the same gas price are generally processed in a first in first out way by Ethereum miners. A contract currently has few options for setting a max gas price:
+
+- Set a constant value at a reasonable rate given the current gas prices
+- Allow an individual or group of individuals to adjust a max gas price over time
+
+More elaborate schemes could likely be constructed but all would involve storing gas price information on chain increasing the number of transactions and costing Ether.
+
+Given a median gas price opcode a contract can set a maximum gas price as a function of the last blocks gas price. This can easily be implemented using a strategy such as the following:
+
+```
+// Assume that block.medgasprice is bound to MEDGASPRICE (0x46)
+
+function submitFraudProof(bytes calldata proof) public {
+  require(tx.gasprice <= maxGasPrice());
+  // process the fraud proof and provide a reward (if valid)
+}
+
+function maxGasPrice() public view returns (uint) {
+  return 3 * block.medgasprice;
+}
+```
+
+Given the contract implementation above a client would simply call `maxGasPrice` to determine the gas price to use when submitting a fraud proof. This particular implementation allows up to 3x the median gas price of the last block to be used.
```

Comment:
This is a use case, which is appropriate for `Motivation` section, not `Rationale` section.
```suggestion
```

Reply:
I added simple reasoning to the rationale section but I'm concerned it's still more motivation than rationale. Let me know if I'm missing it.